src/Lights/Shadows/shadowGenerator.ts

```diff
@@ -776,15 +926,30 @@ export class ShadowGenerator implements IShadowGenerator {
      * @param mapSize The size of the texture what stores the shadows. Example : 1024.
      * @param light The light object generating the shadows.
      * @param usefulFloatFirst By default the generator will try to use half float textures but if you need precision (for self shadowing for instance), you can use this option to enforce full float texture.
+	 * @param cascades Number of cascades for CSM. Values between MIN_CASCADES and MAX_CASCADES.
+	 * @param csmFrustumLength The distance of the furthest frustum of the camera for CSM.
      */
-    constructor(mapSize: number, light: IShadowLight, usefulFloatFirst?: boolean) {
+    constructor(mapSize: number, light: IShadowLight, usefulFloatFirst?: boolean, cascades?: number, csmFrustumLength?: number) {

```

Comment:
not sure the arguments should cotains csm options as it is not the case for the rest of the shadow parameters.

Reply:
So this is the part where I initially considered to code as:
```
Line1: var shadowgenerator = new ShadowGenerator(1024, light);
Line2: shadowgenerator.useCSM = true;
shadowgenerator.<insert filters blah blah>
```
I think that Line1 would create a shadowMap immediately even if the user did not have the intention to use it. And then the useCSM would have to dispose it before initializing csm RTTs. Basically, wasted CPU cycles. Then I thought we could just treat Line1 as a special case of cascade = 1 but I found it a hassle to have to track/rename and reconfigure said map. So I decided to go with:
```
Default
var shadowgenerator = new ShadowGenerator(1024, light);
shadowgenerator.<insert filters blah blah>

CSM
var shadowgenerator = new ShadowGenerator(1024, light, false, 4, 100);
shadowgenerator.<insert filters blah blah>
```
This is very much open for discussion as things could go either way and the difference is minimal. From user perspective. bjs is offering two modes of the basic shadows and it doesn't break back compat. I recognize that it does raise other concerns, biggest of which is whether there will be other types of basic shadows in future to be introduced.

Alternatively, we could just leave it for this PR until its clear that filters or other CSM related improvements do require a new CascadedShadowGenerator class or somesuch...

Reply:
I kind of like the idea of the separate class to be honest as it sounds you have "if CSM" everywhere in this one. It might also be much easier to handle the specific parameter and we could easily limit the mode to Simple and PCF for instance.

What do you think ???

Reply:
I needed to get this first pass out and into people's hand so as to
a) get feedback, bug reports etc
b) get more use case scenarios
c) get an idea of demand and what params users want exposed for tweaking
If you look closely at the csmtransformmatrix calcs, there are major differences compared to the default case. CSM ignores position of the directional light, the lookat transform has no margins atm, casters and receivers are identical for all n maps etc... Plus based on the refs thrown at me, only pcf and vsm are the filtering techniques, csm don't seem to have the same artifact problems plaguing the usual case. I had a sneaky suspicion that it *might* be better off as a separate class. But how to get there ?

Hence, 'if CSM' came about. If CSM users could at least 'play' with it using existing function calls, we would have a route by which a), b) and c) could be fulfilled. That would pave the way for a future CascadedShadowGenerator class. 

To summarize, I have no problems with a new class, but my main concern right now is to get more testing and feedback.

Reply:
We won t be able to push it before 4.2 as we are already in beta and it is a huge change. At least in a separate class this would isolate the changes and leave more room for improvements. I can propose to extract your code as a separate class during the Christmas vacations.

Agree with the filtering technique and I would only consider PCF and PCSS as it might be easier.